### PR TITLE
Cumulative Incidence (Kaplan-Meier) for competing risks.

### LINKED ISSUE #490: Add Kaplan-Meier estimator for competing risks.
**Is your feature request related to a problem? Please describe.**
I would like to include a Kaplan-Meier estimator (cumulative incidences) for the case of competing risks.

**Describe the solution you'd like**
I have already written Python code for it. I am able to reproduce the results of [Scrucca et al.](https://pubmed.ncbi.nlm.nih.gov/17563735/) that follow the cmprsk R package. However, I have still some discrepancies for the confidence interval calculations.
I could send a pull request depending on how serious you consider the issue with the CI. (The differences are 5% for the test data set from Scrucca.)

**Describe alternatives you've considered**
The differences with the cmprsk implementation come from the variance estimators. I have actually implemented them using two formulas (Scrucca papers above) and the equivalent from [Choudhury](https://onlinelibrary.wiley.com/doi/10.1002/sim.1070) and get the same results, but I'm still unable to reproduce the results from the Fortran code in cmprsk.

**References and existing implementations**
https://luca-scr.github.io/software/#competing-risks-analysis
https://cran.r-project.org/web/packages/cmprsk/index.html

### PR DESCRIPTION
**Pull request by mvlvrd**
Add non-parametric Cumulative Incidence (Kaplan-Meier) for competing risks. Tests included.

**Checklist**

- [x] closes #490
- [x] pytest passes
- [x] tests are included
- [x] code is well formatted
- [x] documentation renders correctly

**What does this implement/fix? Explain your changes**
Implements cumulative incidence estimator for the competing risks case. Confidence Intervals are not implemented but are under development.

### CODE REVIEW

**Comment by mvlvrd** [comment]
The codacy linter triggers for unused unpacked assignments, however the ruff linter doesn't. Perhaps it would be nice to make them consistent, though is not a big issue. (I can open a new issue/PR if needed, it is trivial to fix it from the ruff side).

**Review by sebp — CHANGES_REQUESTED** [review]
This is great, thanks a lot!

The code seems to be correct to me. I would suggest to improve the API docs a little bit to avoid ambiguities.

In addition, I'm not sure if certain corner cases are covered. For instance, what happens if no event has been observed for one (or more) competing risk?

I would appreciate if a test case with more than 2 competing risks could be added.

**Comment by mvlvrd** [comment]
Thanks so much for the review!
I think I have addressed most of your points. I have been looking for a real dataset with more than two risks, but I couldn't find one, so I added a simulated one. Corner cases like zero events for one (or more) particular risk are not included and raise an error.

I also changed the function name (removing the reference to Kaplan-Meier). This is because I found a couple of papers that use KM for the naive extension of KM to the competing risks case, while we are using the rigorous estimator that does not assume independence between risks.

File: sksurv/nonparametric.py [hunk 1]
```diff
@@ -586,3 +599,117 @@ def predict_ipcw(self, y):
...
+    Examples
+    --------
+    Creating a Kaplan-Meier curve:
+
+    >>> x, y = km_cumulative_incidence_competing_risks(event, time)
+    >>> for i=range(1, n_risks + 1):
+    >>>    plt.step(x, y[i], where="post", label=i)
+    >>> plt.ylim(0, 1)
+    >>> plt.legend()
+    >>> plt.show()
```
[the remark this answers is not in the document]
Comment (sebp) [reply sksurv/nonparametric.py]: CSV if fine. Can you please add an entry to https://github.com/sebp/scikit-survival/blob/master/sksurv/datasets/data/README.md describing the dataset with a link to its source. Adding a function to load the dataset to `sksurv.datasets` would be great.

File: tests/data/bmt.csv
Comment (sebp) [inline tests/data/bmt.csv:1]: Can you please remove this file and use sksurv.datasets.load_bmt instead

**Review by sebp — CHANGES_REQUESTED** [review]

**Review by sebp — APPROVED** [review]
